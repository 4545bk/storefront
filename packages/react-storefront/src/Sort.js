/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React, { Component } from 'react'
import { withStyles } from '@material-ui/core/'
import { inject, observer } from 'mobx-react'
import classnames from 'classnames'
import Button from '@material-ui/core/Button'
import PropTypes from 'prop-types'

/**
 * UI for sorting an instance of SearchResultModelBase.  This component can be used on its own, or you can use
 * SortButton to automatically display this component in a drawer that slides up from the bottom of the viewport.
 */
export const styles = theme => ({
  container: {
    padding: '15px 0 0 15px',
  },
  option: {
    boxShadow: 'none',
    width: 'calc(50% - 15px)',
    margin: '0 15px 15px 0',
  }
});

@withStyles(styles, { name: 'RSFSort' })
@inject('router')
@observer
export default class Sort extends Component {
  
  static propTypes = {
    /**
     * A function to call when a sort option is selected.  The option and event are passed.
     * The default behavior can be prevented by called `preventDefault()` on the passed in event.
     */
    onSelect: PropTypes.func,

    /**
     * A instance of SearchResultsModelBase
     */
    model: PropTypes.object,

    /**
     * The query string parameter that should be updated when the sort is changed.  The value will the
     * code corresponding to the selected sortOption.  Defaults to "sort".
     */
    queryParam: PropTypes.string
  }

  static defaultProps = {
    onSelect: Function.prototype,
    queryParam: 'sort'
  }

  render() {
    const { model, classes } = this.props
    const options = model.sortOptions

    return (
      <div className={classes.container} onClick={this.handleSort}>
        {options && options.map((option, i) => (
          <Button
            className={classes.option}
            color={model.sort === option.code ? 'primary' : 'default'}
            variant="raised"
            onClick={this.onClick.bind(this, option)}
            key={i}
          >
            {option.name}
          </Button>
        ))}
      </div>
    )
  }

  onClick = (option, e) => {
    this.props.onSelect(option, e)

    if (!e.isDefaultPrevented()) {
      this.props.router.applySearch({
        [this.props.queryParam]: option.code
      })
    }
  }

}
