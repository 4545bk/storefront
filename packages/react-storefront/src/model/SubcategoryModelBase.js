/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import { types } from "mobx-state-tree"
import SearchResultsModelBase from './SearchResultsModelBase'

const SubcategoryModelBase = types.compose(SearchResultsModelBase, 
  types.model("SubcategoryModelBase", {
    id: types.identifier(types.string),
    url: types.maybe(types.string),
    name: types.maybe(types.string),
    description: types.maybe(types.string)
  })
  .views(self => ({
    shouldApplyPatchOnPop(patch) {
      return patch.page === 'Subcategory'
    }
  }))
)

export default SubcategoryModelBase
