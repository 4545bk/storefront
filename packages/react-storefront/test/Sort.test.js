/**
 * @license
 * Copyright Â© 2017-2018 Moov Corporation.  All rights reserved.
 */
import React from 'react'
import { mount } from 'enzyme'
import SubcategoryModelBase from '../src/model/SubcategoryModelBase'
import Sort from '../src/Sort'
import { Provider } from 'mobx-react'

function createModel(sort) {
  return SubcategoryModelBase.create({
    id: '1',
    sort,
    sortOptions: [
      { name: 'Price - Low to High', code: 'price-asc' },
      { name: 'Price - High to Low', code: 'price-desc' },
    ]
  })
}

describe('Sort', () => {

  it('renders with no sort selected', () => {
    expect(mount(
      <Provider router={{}}>
        <Sort model={createModel()}/>
      </Provider>
    )).toMatchSnapshot()
  })

  it('renders with a sort selected', () => {
    expect(mount(
      <Provider router={{}}>
        <Sort model={createModel('price-asc')}/>
      </Provider>
    )).toMatchSnapshot()
  })

})