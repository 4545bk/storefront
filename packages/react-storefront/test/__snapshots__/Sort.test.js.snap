// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort renders with a sort selected 1`] = `
<Provider router={{...}}>
  <WithStyles(inject-Sort-with-router) model={{...}}>
    <inject-Sort-with-router classes={{...}} model={{...}}>
      <Sort classes={{...}} model={{...}} router={{...}} onSelect={[Function]} queryParam="sort">
        <div className="RSFSort-container-1" onClick={[undefined]}>
          <WithStyles(Button) className="RSFSort-option-2" color="primary" variant="raised" onClick={[Function: bound ]}>
            <Button classes={{...}} className="RSFSort-option-2" color="primary" variant="raised" onClick={[Function: bound ]} disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size="medium" type="button">
              <WithStyles(ButtonBase) className="MuiButton-root-3 MuiButton-raised-9 MuiButton-raisedPrimary-10 RSFSort-option-2" disabled={false} focusRipple={true} focusVisibleClassName="MuiButton-focusVisible-12" onClick={[Function: bound ]} type="button">
                <ButtonBase classes={{...}} className="MuiButton-root-3 MuiButton-raised-9 MuiButton-raisedPrimary-10 RSFSort-option-2" disabled={false} focusRipple={true} focusVisibleClassName="MuiButton-focusVisible-12" onClick={[Function: bound ]} type="button" centerRipple={false} disableRipple={false} tabIndex="0">
                  <button onBlur={[Function: handleEvent]} onFocus={[Function: value]} onKeyDown={[Function: value]} onKeyUp={[Function: value]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex="0" className="MuiButtonBase-root-19 MuiButton-root-3 MuiButton-raised-9 MuiButton-raisedPrimary-10 RSFSort-option-2" type="button" disabled={false} onClick={[Function: bound ]}>
                    <span className="MuiButton-label-4">
                      Price - Low to High
                    </span>
                    <WithStyles(TouchRipple) innerRef={[Function: value]} center={false}>
                      <TouchRipple classes={{...}} center={false}>
                        <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-22" childFactory={[Function: childFactory]}>
                          <span className="MuiTouchRipple-root-22" />
                        </TransitionGroup>
                      </TouchRipple>
                    </WithStyles(TouchRipple)>
                  </button>
                </ButtonBase>
              </WithStyles(ButtonBase)>
            </Button>
          </WithStyles(Button)>
          <WithStyles(Button) className="RSFSort-option-2" color="default" variant="raised" onClick={[Function: bound ]}>
            <Button classes={{...}} className="RSFSort-option-2" color="default" variant="raised" onClick={[Function: bound ]} disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size="medium" type="button">
              <WithStyles(ButtonBase) className="MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" disabled={false} focusRipple={true} focusVisibleClassName="MuiButton-focusVisible-12" onClick={[Function: bound ]} type="button">
                <ButtonBase classes={{...}} className="MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" disabled={false} focusRipple={true} focusVisibleClassName="MuiButton-focusVisible-12" onClick={[Function: bound ]} type="button" centerRipple={false} disableRipple={false} tabIndex="0">
                  <button onBlur={[Function: handleEvent]} onFocus={[Function: value]} onKeyDown={[Function: value]} onKeyUp={[Function: value]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex="0" className="MuiButtonBase-root-19 MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" type="button" disabled={false} onClick={[Function: bound ]}>
                    <span className="MuiButton-label-4">
                      Price - High to Low
                    </span>
                    <WithStyles(TouchRipple) innerRef={[Function: value]} center={false}>
                      <TouchRipple classes={{...}} center={false}>
                        <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-22" childFactory={[Function: childFactory]}>
                          <span className="MuiTouchRipple-root-22" />
                        </TransitionGroup>
                      </TouchRipple>
                    </WithStyles(TouchRipple)>
                  </button>
                </ButtonBase>
              </WithStyles(ButtonBase)>
            </Button>
          </WithStyles(Button)>
        </div>
      </Sort>
    </inject-Sort-with-router>
  </WithStyles(inject-Sort-with-router)>
</Provider>
`;

exports[`Sort renders with no sort selected 1`] = `
<Provider router={{...}}>
  <WithStyles(inject-Sort-with-router) model={{...}}>
    <inject-Sort-with-router classes={{...}} model={{...}}>
      <Sort classes={{...}} model={{...}} router={{...}} onSelect={[Function]} queryParam="sort">
        <div className="RSFSort-container-1" onClick={[undefined]}>
          <WithStyles(Button) className="RSFSort-option-2" color="default" variant="raised" onClick={[Function: bound ]}>
            <Button classes={{...}} className="RSFSort-option-2" color="default" variant="raised" onClick={[Function: bound ]} disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size="medium" type="button">
              <WithStyles(ButtonBase) className="MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" disabled={false} focusRipple={true} focusVisibleClassName="MuiButton-focusVisible-12" onClick={[Function: bound ]} type="button">
                <ButtonBase classes={{...}} className="MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" disabled={false} focusRipple={true} focusVisibleClassName="MuiButton-focusVisible-12" onClick={[Function: bound ]} type="button" centerRipple={false} disableRipple={false} tabIndex="0">
                  <button onBlur={[Function: handleEvent]} onFocus={[Function: value]} onKeyDown={[Function: value]} onKeyUp={[Function: value]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex="0" className="MuiButtonBase-root-19 MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" type="button" disabled={false} onClick={[Function: bound ]}>
                    <span className="MuiButton-label-4">
                      Price - Low to High
                    </span>
                    <WithStyles(TouchRipple) innerRef={[Function: value]} center={false}>
                      <TouchRipple classes={{...}} center={false}>
                        <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-22" childFactory={[Function: childFactory]}>
                          <span className="MuiTouchRipple-root-22" />
                        </TransitionGroup>
                      </TouchRipple>
                    </WithStyles(TouchRipple)>
                  </button>
                </ButtonBase>
              </WithStyles(ButtonBase)>
            </Button>
          </WithStyles(Button)>
          <WithStyles(Button) className="RSFSort-option-2" color="default" variant="raised" onClick={[Function: bound ]}>
            <Button classes={{...}} className="RSFSort-option-2" color="default" variant="raised" onClick={[Function: bound ]} disabled={false} disableFocusRipple={false} fullWidth={false} mini={false} size="medium" type="button">
              <WithStyles(ButtonBase) className="MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" disabled={false} focusRipple={true} focusVisibleClassName="MuiButton-focusVisible-12" onClick={[Function: bound ]} type="button">
                <ButtonBase classes={{...}} className="MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" disabled={false} focusRipple={true} focusVisibleClassName="MuiButton-focusVisible-12" onClick={[Function: bound ]} type="button" centerRipple={false} disableRipple={false} tabIndex="0">
                  <button onBlur={[Function: handleEvent]} onFocus={[Function: value]} onKeyDown={[Function: value]} onKeyUp={[Function: value]} onMouseDown={[Function: handleEvent]} onMouseLeave={[Function: handleEvent]} onMouseUp={[Function: handleEvent]} onTouchEnd={[Function: handleEvent]} onTouchMove={[Function: handleEvent]} onTouchStart={[Function: handleEvent]} tabIndex="0" className="MuiButtonBase-root-19 MuiButton-root-3 MuiButton-raised-9 RSFSort-option-2" type="button" disabled={false} onClick={[Function: bound ]}>
                    <span className="MuiButton-label-4">
                      Price - High to Low
                    </span>
                    <WithStyles(TouchRipple) innerRef={[Function: value]} center={false}>
                      <TouchRipple classes={{...}} center={false}>
                        <TransitionGroup component="span" enter={true} exit={true} className="MuiTouchRipple-root-22" childFactory={[Function: childFactory]}>
                          <span className="MuiTouchRipple-root-22" />
                        </TransitionGroup>
                      </TouchRipple>
                    </WithStyles(TouchRipple)>
                  </button>
                </ButtonBase>
              </WithStyles(ButtonBase)>
            </Button>
          </WithStyles(Button)>
        </div>
      </Sort>
    </inject-Sort-with-router>
  </WithStyles(inject-Sort-with-router)>
</Provider>
`;
